### PR TITLE
.github/workflows: add CI on IoT-LAB

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds a github action workflow that launch `compile_and_test_for_board` on a list of boards provided by the IoT-LAB testbed.
For each board, an experiment is launched, then the compile_and_test script starts. Once it completes, the experiment is stopped and build artifacts are stored. These 2 steps are always performed, regardless of the compile and test status.

For each board, jobs are run in parallel and a job takes ~2h30 to complete. Only 6 boards are added for the moment but more could be added (samr30-xpro, frdm-kw41z, nrf52dk, nrf52840dk, etc).

Since it takes a lot of time to complete, it's not ~possible~ acceptable to perform this check on PR basis, so the action is configured to be run on a weekly basis on master and for each RC tag pushed.

This action uses a specific account on IoT-LAB and to be able to launch an experiment and interact with the serial port 2 secrets must be configured on the GitHub project (@miri64 knows about that already). I can send the SSH key via email or chat.

### Testing procedure

See [here](https://github.com/aabadie/RIOT/runs/901811217?check_suite_focus=true) to see what it looks like.

### Issues/PRs references

None

### CODE REVIEW

**Comment by aabadie** [comment]
One thing I forgot to mention is that on IoT-LAB it's not (yet) possible to flash a firmware at a given offset: so all related applications (suite_update, riotboot, suit_manifest) are failing. Maybe we could skip these using the `--applications-exclude` option of `compile_and_test_for_board.py` until this is supported in iot-lab.

File: .github/workflows/iotlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  # Schedule weekly runs Sunday at 00:00 on master
+  schedule:
+    - cron: '00 0 * * 0'
+  push:
+    # TODO: remove master branch, only used here as demonstrator
+    branches:
+      - master
+    # Run on all new release candidates tags
+    tags:
+      - '*-RC*'
```
Comment (miri64) [inline .github/workflows/iotlab-ci.yml:20]: I think we can be a bit more strict here
```suggestion
      - '[0-9][0-9][0-9][0-9]-RC[0-9]*'
```

File: .github/workflows/iotlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+jobs:
+  # Runs all tests on IoT-LAB boards
+  compile_and_test_for_board:
+    strategy:
+      max-parallel: 10
```
Comment (miri64) [inline .github/workflows/iotlab-ci.yml:26]:
```suggestion
      max-parallel: 6
```

?

File: .github/workflows/iotlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      fail-fast: false
+      matrix:
+        boards:
+          - '{"riot": "iotlab-m3", "iotlab": "m3:at86rf231", "site": "saclay"}'
+          - '{"riot": "nrf51dk", "iotlab": "nrf51dk:ble", "site": "saclay"}'
+          - '{"riot": "nrf52832-mdk", "iotlab": "nrf52832mdk:ble", "site": "saclay"}'
+          - '{"riot": "samr21-xpro", "iotlab": "samr21:at86rf233", "site": "saclay"}'
+          - '{"riot": "b-l072z-lrwan1", "iotlab": "st-lrwan1:sx1276", "site": "saclay"}'
+          - '{"riot": "b-l475e-iot01a", "iotlab": "st-iotnode:multi", "site": "saclay"}'
```
Comment (miri64) [inline .github/workflows/iotlab-ci.yml:35]: Can't this be `yaml`? This way the usage of `fromJson` below wouldn't be needed.
```suggestion
          - riot: iotlab-m3
            iotlab: "m3:at86rf231"
            site: saclay
          - riot: nrf51dk
            iotlab: "nrf51dk:ble"
            site: saclay
          - riot: nrf52832-mdk
            iotlab: "nrf52832mdk:ble"
            site: saclay
          - riot: samr21-xpro
            iotlab: "samr21:at86rf233"
            site: saclay
          - riot: b-l072z-lrwan1
            iotlab: "st-lrwan1:sx1276"
            site: saclay
          - riot: b-l475e-iot01a
            iotlab: "st-iotnode:multi"
            site: saclay
```

File: .github/workflows/iotlab-ci.yml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+jobs:
+  # Runs all tests on IoT-LAB boards
+  compile_and_test_for_board:
+    strategy:
+      max-parallel: 10
+      fail-fast: false
+      matrix:
+        boards:
+          - '{"riot": "iotlab-m3", "iotlab": "m3:at86rf231", "site": "saclay"}'
```
Comment (miri64) [inline .github/workflows/iotlab-ci.yml:30]: Maybe also do something like this?

```suggestion
          - riot: iotlab-m3
            iotlab:
              archi: "m3:at86rf231"
              site: saclay
```

**Comment by miri64** [comment]
> This action uses a specific account on IoT-LAB and to be able to launch an experiment and interact with the serial port 2 secrets must be configured on the GitHub project (@miri64 knows about that already). I can send the SSH key via email or chat.

Would it make sense, that we use the same SSH key?

**Comment by aabadie** [comment]
> Would it make sense, that we use the same SSH key?

Yes, and it'll be simpler since you already have it.

**Comment by miri64** [comment]
> > Would it make sense, that we use the same SSH key?
> 
> Yes, and it'll be simpler since you already have it.

I added it to the RIOT-OS org with permissions to be used here :-).

**Comment by aabadie** [comment]
Other comments addressed.

**Comment by aabadie** [comment]
And added nrf52dk and nrf52840dk boards.

**Comment by miri64** [comment]
I think you need to kick it so we can see if it works ;-)

**Comment by miri64** [comment]
(I would add a `workflow_dispatch` trigger for that)

**Comment by aabadie** [comment]
> I would add a workflow_dispatch trigger for that

Added and [seems to work](https://github.com/aabadie/RIOT/actions/runs/188207662)

File: .github/workflows/iotlab-ci.yml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+# pushed release candidates tags.
+
+on:
+  # Schedule weekly runs on Monday at 00:00 on master
+  schedule:
+    - cron: '00 0 * * 1'
+  push:
+    # Run on all new release candidates tags
+    tags:
+      - '[0-9][0-9][0-9][0-9]-RC[0-9]*'
```
Comment (miri64) [inline .github/workflows/iotlab-ci.yml:17]: Oops, there is now the month missing :sweat_smile: 

```suggestion
      - '[0-9][0-9][0-9][0-9].[0-9][0-9]-RC[0-9]*'
```

Comment (miri64) [reply .github/workflows/iotlab-ci.yml:17]: Also does it make sense to also build for releases and point releases?

```suggestion
      - '[0-9][0-9][0-9][0-9].[0-9][0-9]-RC[0-9]*'
      - '[0-9][0-9][0-9][0-9].[0-9][0-9]'
      - '[0-9][0-9][0-9][0-9].[0-9][0-9].*'
```

Comment (aabadie) [reply .github/workflows/iotlab-ci.yml]: Fixed

**Comment by miri64** [comment]
Can you give this a final trigger?

File: .github/workflows/iotlab-ci.yml [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  schedule:
+    - cron: '00 0 * * 1'
+  push:
+    # Run on all new releases and release candidates tags
+    tags:
+      - '[0-9][0-9][0-9][0-9].[0-9][0-9]-RC[0-9]*'
+      - '[0-9][0-9][0-9][0-9].[0-9][0-9]'
+      - '[0-9][0-9][0-9][0-9].[0-9][0-9].*'
+  # Allow manual trigger
+  workflow_dispatch:
```
Comment (miri64) [inline .github/workflows/iotlab-ci.yml:21]:
```suggestion
  workflow_dispatch:
    inputs:
      riot_version:
        description: 'RIOT version to checkout'
        required: true
        default: 'master'
```

File: .github/workflows/iotlab-ci.yml [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      - name: Fetch host key from IoT-LAB ${{ matrix.boards.iotlab.site }} site
+        run: |
+          IOTLAB_USER=$(cat ~/.iotlabrc | cut -f1 -d:)
+          ssh -oStrictHostKeyChecking=accept-new \
+          ${IOTLAB_USER}@${{ matrix.boards.iotlab.site }}.iot-lab.info exit
+      - name: Pull riotbuild docker image
+        run: |
+          docker pull riot/riotbuild:latest
+      - name: Checkout RIOT
+        uses: actions/checkout@v2
```
Comment (miri64) [inline .github/workflows/iotlab-ci.yml:98]:
```suggestion
        uses: actions/checkout@v2
        ref: ${{ github.event.inputs.riot_version }}
```

File: .github/workflows/iotlab-ci.yml [hunk 8]
```diff
@@ -0,0 +1,119 @@
+name: iotlab-ci
```
Comment (miri64) [inline .github/workflows/iotlab-ci.yml:1]: Also, I'm not sure about the name... how about (not sure I like that either, but it is more descriptive...)
```suggestion
name: compile-and-test-in-iotlab
```

Comment (miri64) [reply .github/workflows/iotlab-ci.yml:1]: Its just: The current name sticks out as very non-descriptive

image

Comment (aabadie) [reply .github/workflows/iotlab-ci.yml:1]: My personal preference goes for the original name: it's descriptive enough (CI on IoT-LAB) and short.

Comment (miri64) [reply .github/workflows/iotlab-ci.yml:1]: >  it's descriptive enough (CI on IoT-LAB) and short.

What is CI on `iotlab`? The release tests are also CI. Running a specific on `iotlab-m3` is also `iotlab-ci`. You see that it is not very specific?

Comment (aabadie) [reply .github/workflows/iotlab-ci.yml:1]: > What is CI on iotlab? 

Running the tests on boards provided by IoT-LAB. As usual, we can nitpicks on words and loose our time. I choose this names for a good reason, I prefer to keep it.

Comment (miri64) [reply .github/workflows/iotlab-ci.yml:1]: > Running the tests on boards provided by IoT-LAB. […]

Then maybe `test-on-iotlab`?

> As usual, we can nitpicks on words and loose our time. I choose this names for a good reason, I prefer to keep it.

I don't think I am nitpicking in this instance. Names are important, especially if they are shown very prominently. I don't think the name `iotlab-ci` is not very good.

Comment (aabadie) [reply .github/workflows/iotlab-ci.yml:1]: > I don't think the name iotlab-ci is not very good.

So you think it's good ? So do I :)

Comment (miri64) [reply .github/workflows/iotlab-ci.yml:1]: > > I don't think the name iotlab-ci is not very good.
> 
> So you think it's good ? So do I :)

:sweat_smile: ... fine let's not delay this PR on something that can be changed later :smiling_imp: :stuck_out_tongue_winking_eye:

Comment (fjmolinas) [reply .github/workflows/iotlab-ci.yml:1]: I agree with @miri64 on the name not being the most descriptive. I would like to see the word `tests` or `run-tests` in the name as well.

**Comment by aabadie** [comment]
> Do you have a comment on #14587 (review)?

I'm testing it on my fork.

**Comment by miri64** [comment]
> > Do you have a comment on #14587 (review)?
> 
> I'm testing it on my fork.

You need to configure the branch with this YAML as default branch to get the input mask in actions.

File: .github/workflows/iotlab-ci.yml (hunk 7 above)
Comment (aabadie) [reply .github/workflows/iotlab-ci.yml:98]: I don't see how this would work on pushes of RC or release tags.

Comment (miri64) [reply .github/workflows/iotlab-ci.yml:98]: In that case it is an empty string in which case the ref under test is pulled. See https://github.com/actions/checkout#usage

File: .github/workflows/test-on-iotlab.yml [hunk 9]
```diff
@@ -0,0 +1,129 @@
+name: test-on-iotlab.yml
```
Comment (miri64) [inline .github/workflows/test-on-iotlab.yml:1]:
```suggestion
name: test-on-iotlab
```
?

File: .github/workflows/test-on-iotlab.yml [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+      - name: Pull riotbuild docker image
+        run: |
+          docker pull riot/riotbuild:latest
+      - name: Checkout RIOT
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event.inputs.riot_version }}
+      - name: Print RIOT version
+        run: |
+          git log -1 --pretty
```
Comment (miri64) [inline .github/workflows/test-on-iotlab.yml:108]: The [checkout action does this already for you](https://github.com/aabadie/RIOT/runs/926488305?check_suite_focus=true#step:8:457) though not as pretty ;-). Still, I think this information is redundant.

**Comment by aabadie** [comment]
It should be good now:
- triggering a workflow from the hash of the last commit of a PR works
- auto trigger of pushed tags works
- for completeness, we can wait monday morning to see if the cron triggers the right commit

And the workflow is now called `test-on-iotlab`, the yaml file is called the same.

File: .github/workflows/test-on-iotlab.yml (hunk 9 above)
Comment (aabadie) [reply .github/workflows/test-on-iotlab.yml]: Oups, fixed!

File: .github/workflows/test-on-iotlab.yml (hunk 10 above)
Comment (aabadie) [reply .github/workflows/test-on-iotlab.yml]: Left-over from my previous tests...

**Comment by miri64** [comment]
Manually started [the first run](https://github.com/RIOT-OS/RIOT/actions/runs/190219301). Let's see, if I configured everything correct in the orga :-)

**Comment by miri64** [comment]
> Manually started [the first run](https://github.com/RIOT-OS/RIOT/actions/runs/190219301). Let's see, if I configured everything correct in the orga :-)

Yepp ran all the way through (but failed of course for some.

**Comment by aabadie** [comment]
Thanks for reviewing and merging!

**Comment by miri64** [comment]
You're welcome! We should not forget the main event though ;-) #14646

**Comment by miri64** [comment]
The tests ran as scheduled btw :-) https://github.com/RIOT-OS/RIOT/actions/runs/192612232

**Pull request merged**